File: packages/gavio-js/src/interceptors/governance/drift.ts
/**
 * driftMonitor (F-GOV-07) — alert when a provider's response distribution shifts.
 *
 * A `DriftDetector` is fed one metric sample per request (latency, tokens, …);
 * the default `StatisticalDriftDetector` keeps a rolling-window baseline and
 * flags a sample that deviates beyond a z-score threshold. Alerts surface as
 * `governance.event` inspector events (and in `/api/stats`) and are logged.
 */

import type { InterceptorContext } from '../../context.js'
import type { GavioResponse } from '../../response.js'
import type { Interceptor } from '../base.js'

export interface DriftBaseline {
  mean: number
  std: number
  n: number
}

export interface DriftAlert {
  metric: string
  value: number
  baseline: DriftBaseline
  /** Standard scores from the baseline mean; null when the baseline had zero variance. */
  z: number | null
  threshold: number
}

/** Pluggable drift detector: fed per-request samples, returns any alerts. */
export interface DriftDetector {
  readonly name: string
  observe(sample: Record<string, number>): DriftAlert[]
}

function round(value: number, decimals = 4): number {
  const factor = 10 ** decimals
  return Math.round(value * factor) / factor
}

function meanStd(values: number[]): { mean: number; std: number } {
  const n = values.length
  const mean = values.reduce((a, b) => a + b, 0) / n
  const variance = values.reduce((a, b) => a + (b - mean) ** 2, 0) / n
  return { mean, std: Math.sqrt(variance) }
}

export interface StatisticalDriftDetectorOptions {
  /** Rolling baseline window size (default 50). */
  windowSize?: number
  /** Minimum samples before any alert fires (default = windowSize). */
  minSamples?: number
  /** z-score sensitivity — deviation beyond this many std devs alerts (default 3). */
  threshold?: number
}

/** Rolling-window z-score detector — the default `DriftDetector`. */
export class StatisticalDriftDetector implements DriftDetector {
  readonly name = 'statistical'
  private readonly windowSize: number
  private readonly minSamples: number
  private readonly threshold: number
  private readonly windows = new Map<string, number[]>()

  constructor(options: StatisticalDriftDetectorOptions = {}) {
    this.windowSize = options.windowSize ?? 50
    this.minSamples = options.minSamples ?? this.windowSize
    this.threshold = options.threshold ?? 3
  }

  observe(sample: Record<string, number>): DriftAlert[] {
    const alerts: DriftAlert[] = []
    for (const [metric, value] of Object.entries(sample)) {
      if (!Number.isFinite(value)) continue
      let window = this.windows.get(metric)
      if (window === undefined) {
        window = []
        this.windows.set(metric, window)
      }
      if (window.length >= this.minSamples) {
        const { mean, std } = meanStd(window)
        const baseline: DriftBaseline = { mean: round(mean), std: round(std), n: window.length }
        if (std > 0) {
          const z = (value - mean) / std
          if (Math.abs(z) > this.threshold) {
            alerts.push({ metric, value, baseline, z: round(z), threshold: this.threshold })
          }
        } else if (value !== mean) {
          alerts.push({ metric, value, baseline, z: null, threshold: this.threshold })
        }
      }
      window.push(value)
      if (window.length > this.windowSize) window.shift()
    }
    return alerts
  }
}

export interface DriftMonitorOptions extends StatisticalDriftDetectorOptions {
  /** Metrics to watch (default `['latency_ms', 'total_tokens']`). */
  metrics?: string[]
  /** Override the detector (default a `StatisticalDriftDetector`). */
  detector?: DriftDetector
}

function extractSample(
  metrics: string[],
  response: GavioResponse,
  ctx: InterceptorContext,
): Record<string, number> {
  const sample: Record<string, number> = {}
  for (const metric of metrics) {
    if (metric === 'latency_ms') sample[metric] = response.latencyMs
    else if (metric === 'total_tokens') sample[metric] = response.usage.totalTokens
    else if (metric === 'cost_usd') sample[metric] = response.costUsd
    else if (metric === 'risk_score' && ctx.riskScore != null) sample[metric] = ctx.riskScore
  }
  return sample
}

/** Observe-only interceptor that flags response-distribution drift (F-GOV-07). */
export function driftMonitor(options: DriftMonitorOptions = {}): Interceptor {
  const metrics = options.metrics ?? ['latency_ms', 'total_tokens']
  const detector = options.detector ?? new StatisticalDriftDetector(options)
  return {
    name: 'drift_monitor',
    dryRunSafe: false, // never let a dry run pollute the baseline
    after(response: GavioResponse, ctx: InterceptorContext): GavioResponse {
      for (const alert of detector.observe(extractSample(metrics, response, ctx))) {
        ctx.recordGovernanceEvent({
          kind: 'drift',
          detector: detector.name,
          metric: alert.metric,
          value: alert.value,
          baseline: alert.baseline,
          z: alert.z,
          threshold: alert.threshold,
        })
        // eslint-disable-next-line no-console
        console.warn(
          `[gavio:drift] ${alert.metric}=${alert.value} drifted from baseline ` +
            `mean=${alert.baseline.mean} std=${alert.baseline.std} (n=${alert.baseline.n})`,
        )
      }
      return response
    },
  }
}

File: packages/gavio-py/gavio/interceptors/governance/drift.py
"""DriftMonitor (F-GOV-07) — alert when a provider's response distribution shifts.

A ``DriftDetector`` is fed one metric sample per request (latency, tokens, …);
the default ``StatisticalDriftDetector`` keeps a rolling-window baseline and
flags a sample that deviates beyond a z-score threshold. Alerts surface as
``governance.event`` inspector events (and in ``/api/stats``) and are logged.
"""

from __future__ import annotations

import logging
import math
import statistics
from collections import deque
from dataclasses import dataclass
from typing import Any, Protocol, runtime_checkable

from ...context import InterceptorContext
from ...response import GavioResponse
from ..base import Interceptor

logger = logging.getLogger("gavio.drift")


@dataclass
class DriftAlert:
    metric: str
    value: float
    baseline: dict[str, Any]  # {"mean": ..., "std": ..., "n": ...}
    z: float | None  # None when the baseline had zero variance
    threshold: float


@runtime_checkable
class DriftDetector(Protocol):
    """Pluggable drift detector: fed per-request samples, returns any alerts."""

    @property
    def name(self) -> str: ...

    def observe(self, sample: dict[str, float]) -> list[DriftAlert]: ...


class StatisticalDriftDetector:
    """Rolling-window z-score detector — the default ``DriftDetector``."""

    def __init__(
        self,
        window_size: int = 50,
        min_samples: int | None = None,
        threshold: float = 3.0,
    ) -> None:
        self.window_size = window_size
        self.min_samples = min_samples if min_samples is not None else window_size
        self.threshold = threshold
        self._windows: dict[str, deque[float]] = {}

    @property
    def name(self) -> str:
        return "statistical"

    def observe(self, sample: dict[str, float]) -> list[DriftAlert]:
        alerts: list[DriftAlert] = []
        for metric, value in sample.items():
            if not math.isfinite(value):
                continue
            window = self._windows.setdefault(metric, deque())
            if len(window) >= self.min_samples:
                mean = statistics.fmean(window)
                std = statistics.pstdev(window)
                baseline = {"mean": round(mean, 4), "std": round(std, 4), "n": len(window)}
                if std > 0:
                    z = (value - mean) / std
                    if abs(z) > self.threshold:
                        alerts.append(
                            DriftAlert(metric, value, baseline, round(z, 4), self.threshold)
                        )
                elif value != mean:
                    alerts.append(DriftAlert(metric, value, baseline, None, self.threshold))
            window.append(value)
            if len(window) > self.window_size:
                window.popleft()
        return alerts


class DriftMonitor(Interceptor):
    """Observe-only interceptor that flags response-distribution drift (F-GOV-07)."""

    def __init__(
        self,
        metrics: list[str] | None = None,
        detector: DriftDetector | None = None,
        window_size: int = 50,
        min_samples: int | None = None,
        threshold: float = 3.0,
    ) -> None:
        self.metrics = metrics if metrics is not None else ["latency_ms", "total_tokens"]
        self.detector = detector or StatisticalDriftDetector(window_size, min_samples, threshold)

    @property
    def name(self) -> str:
        return "drift_monitor"

    @property
    def dry_run_safe(self) -> bool:
        # Never let a dry run pollute the baseline.
        return False

    def _extract(self, response: GavioResponse, ctx: InterceptorContext) -> dict[str, float]:
        sample: dict[str, float] = {}
        for metric in self.metrics:
            if metric == "latency_ms":
                sample[metric] = response.latency_ms
            elif metric == "total_tokens":
                sample[metric] = response.usage.total_tokens
            elif metric == "cost_usd":
                sample[metric] = response.cost_usd
            elif metric == "risk_score" and ctx.risk_score is not None:
                sample[metric] = ctx.risk_score
        return sample

    async def after(self, response: GavioResponse, ctx: InterceptorContext) -> GavioResponse:
        for alert in self.detector.observe(self._extract(response, ctx)):
            ctx.record_governance_event(
                {
                    "kind": "drift",
                    "detector": self.detector.name,
                    "metric": alert.metric,
                    "value": alert.value,
                    "baseline": alert.baseline,
                    "z": alert.z,
                    "threshold": alert.threshold,
                }
            )
            logger.warning(
                "drift: %s=%s from baseline mean=%s std=%s (n=%s)",
                alert.metric,
                alert.value,
                alert.baseline["mean"],
                alert.baseline["std"],
                alert.baseline["n"],
            )
        return response
